### ISSUE TITLE
Fast Gradient Method do not generate adversarials in some cases

### ISSUE THREAD

**Issue by KevinRupper** [Trusted-AI/adversarial-robustness-toolbox#894]
Hi,

I encounter a strange behavior generating adversarial examples (non-targeted) for MobileNetV2, DenseNet201 or ResNet152V2 nets. Given 100 images from Imagenet selected randomly the algorithm is only able to generate 17 adversarial images to ResNet152V2 net, the rest of the images remains clear without any sign of having been modified (the same happen for DenseNet201 and MobileNetV2). For other architectures (like VGG16, VGG19 and ResNet50) the attack successfully generates 100 images clearly tampered. Since FGM is the most simple attack and only calculate the gradients respect to the input and apply eps value to the results I expected that all images appear modified for all nets even if the perturbations does not achieve to deceive the model.

The configuration of the environment is as follow:

- TensorFlow/Keras: 2.3.1
- ART: 1.5.1
- Models loaded from Keras with Imagenet pre-trained weights
- Attack params: eps = 30 (intentionally high to show the perturbation), the rest of the parameters remains by default
- The images are divided in two batches of 50 images

__init__(estimator: CLASSIFIER_LOSS_GRADIENTS_TYPE, norm: Union[int, float, str] = inf, eps: Uni
on[int, float, numpy.ndarray] = 0.3, eps_step: Union[int, float, numpy.ndarray] = 0.1, targeted: bool = False, num_random_init: int = 0, batch_size: int = 32, minimal: bool = False)

![ILSVRC2012_val_00010618](https://user-images.githubusercontent.com/3661776/107636614-2e5cc980-6c6d-11eb-8472-e16a9a5b4675.JPEG) ![ILSVRC2012_val_00003712](https://user-images.githubusercontent.com/3661776/107636642-39175e80-6c6d-11eb-96e1-6dd957ac0b47.JPEG)

I observed the same behavior for Projected Gradient Descent, but in this case, I can understand that is an iterative method and perhaps the algorithm is not capable to perform adversarials from a more complex architectures with low iterations values.

Perpahs I missing something from the FGM algorithm or from the implementation. Please mark this issue as a question.

Thank you.

**Comment by beat-buesser** [comment]
Hi @KevinRupper Thank you very much for using ART! I would be interested to learn the reason for you observations. Would you be able to share a Jupyter notebook with your code and experiments?

**Comment by KevinRupper** [comment]
Hi @beat-buesser, you can check the code here: https://github.com/KevinRupper/phd

**Comment by beat-buesser** [comment]
Hi @KevinRupper How does your project use ART? Could you please print the loss values and/or predicted probabilities for some of the successful and unsuccessful adversarial examples?

**Comment by KevinRupper** [comment]
@beat-buesser I use ART to perform the attacks based on the configuration params and save each of them, the images were picked randomly from test set of Imagenet dataset but with a threshold of .70 of confidence. Here is the loop where the attack is performed:

```print('Generating {0} adversarials for {1} with params: {2}'.format(attack_type, net.name(), attack_params))
    for idx, batch in enumerate(batches):        
        images = iu.load_batch(batch, dataset, net.input_shape()[0], net.input_shape()[1])
        # Perform the attack
        adversarials = attack.generate(x = images)
        for idx, image in enumerate(adversarials):
            image = image_preprocess.array_to_img(image)
            path = os.path.join(adversarials_folder, batch[idx])
            image.save(path, format = 'JPEG', subsampling = 0, quality = 100)```

After that, I use the models to predict the new adversarial images and compare the label result with the ground truth (I don't have the loss values and predicted values because in the algorithm I only measure if the attack success). To reproduce the experiment you need to load a ResNet152V2 model from Keras:

```net = ResNet152V2(include_top = True, weights = 'imagenet')```

Pick 20 images or so, load them and perform the attack in batch with "inf" norm and eps 30. Later check if the 20 images have been tampered.

**Comment by KevinRupper** [comment]
Hi @beat-buesser, here you have a little notebook to reproduce the experiment:
https://colab.research.google.com/drive/19b6lTiPMWTo0pgTYkpCc20xEDZBAxvwG?usp=sharing

Use this 100 images to test against: [images.zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/6004776/images.zip)

And here more examples of the results:

![Captura de pantalla 2021-02-18 a las 18 46 21](https://user-images.githubusercontent.com/3661776/108398848-c16fa380-7219-11eb-85ba-f9b749829269.png)
![Captura de pantalla 2021-02-18 a las 18 46 33](https://user-images.githubusercontent.com/3661776/108398871-c6ccee00-7219-11eb-92d6-a0631c83ff9e.png)

**Comment by beat-buesser** [comment]
Hi @KevinRupper 
Thank you for the colab notebook, that's very helpful. A few things I have noticed:
- Are you using the correct preprocessing of the images? I didn't see `tf.keras.applications.resnet_v2.preprocess_input` in the notebook
- I think `images` is a list of `np.nadarray`, it should be stacked to a single array of format NHWC

**Comment by KevinRupper** [comment]
Hi @beat-buesser, I have changed the notebook with your suggestions, although I observed some strange results with other nets, only the ResNet152V2, MobileNetV2 and DenseNet201 nets seem to be affected now (the rest probably are affected but with low noise). I don't now if this results are normal and inherent to each net, as I mention above I expected similar noise results for all nets with same "eps" value. Below you can find a table with the result of the attack Fast Gradient (eps=0.1, norm="inf") with both, pre-processed and unprocessed images:

|  Net       | Pre-processed           | Unprocessed  |
| ------------- |:-------------:| -----:|
| Vgg19 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632770-f164b400-7470-11eb-8c16-b953f36ff745.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632776-f9bcef00-7470-11eb-99b4-58c0f4f782ee.JPEG) |
| Vgg16 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632825-2bce5100-7471-11eb-88d3-1c3175f87856.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632828-32f55f00-7471-11eb-9e92-3d8919a4220b.JPEG) |
| ResNet152V2 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632850-499bb600-7471-11eb-8c5a-c9aa3db178a1.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632858-528c8780-7471-11eb-9ee2-ea557b370da3.JPEG) |
| ResNet50 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632876-6932de80-7471-11eb-9b38-dfa72df4df0b.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632882-72bc4680-7471-11eb-9eef-473a25af8573.JPEG) |
| MobileNetV2 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632897-84055300-7471-11eb-87a6-f63d1a349fd4.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632903-8b2c6100-7471-11eb-92bd-27dce9cc4539.JPEG) |
| EfficientNetB0 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632921-a26b4e80-7471-11eb-897c-ec1cc58c81e4.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632931-a9925c80-7471-11eb-957f-e6c43345f89d.JPEG) |
| DenseNet201 | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632941-b911a580-7471-11eb-9e88-74374554a230.JPEG) | ![ILSVRC2012_val_00026733](https://user-images.githubusercontent.com/3661776/108632945-c29b0d80-7471-11eb-8309-4ec613b4019c.JPEG) |

**Comment by beat-buesser** [comment]
Hi @KevinRupper

Sorry, I didn't notice it the first time. You need to provide the un-preprocessed images to `generate` and define the classifier with the `preprocessing` argument:
```
classifier = KerasClassifier(model = net, preprocessing=(mean, std))
```
where the values for `mean` and `std` are from the model-specific preprocessing functions, e.g. https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/applications/imagenet_utils.py#L169

Or you can implement instance of `art.preprocessing` which would wrap the preprocessing function in its method `forward` to automatically apply and backpropagate gradients through it.

The main reason for this is that the preprocessed images and the provided `eps` are not in the same pixel space anymore.

**Comment by KevinRupper** [comment]
Hi @beat-buesser, I have been testing removing the preprocessing step and with `mean` and `std` values of VGG family nets but without success creating the attack, the results of the attack are black images. 

The following chunk of code is extracted from Keras source of `vgg19.py:`

`def preprocess_input(x, data_format=None):
  return imagenet_utils.preprocess_input( x, data_format=data_format, mode='caffe')`

And these are the values for `caffe` type and `None` data format in `imagenet_utils.py`:

`   mean = [103.939, 116.779, 123.68]
    std = None`

I updated the notebook with the preprocessing call:

`classifier = KerasClassifier(model = net, preprocessing = ([103.939, 116.779, 123.68], None))`

**Comment by beat-buesser** [comment]
Hi @KevinRupper
I looked again at the notebook, a few more things I noticed:
- the `images` and the preprocessing with `mean` and `std` have to be in the same pixel value range, for example if `mean` is `[0.485, 0.456, 0.406]` the images should probably be in range 0-1, but currently they seem in range 0-255, and vice versa
- change `adversarials = attack.generate(x = images)` to `adversarials = attack.generate(x=images, y=labels)` to attack the true labels and make the attack stronger. Currently you are attacking the model prediction and if the model is very confident about its prediction the loss will be zero and therefore also the loss gradients required for the attack.

**Comment by KevinRupper** [comment]
Hi @beat-buesser,

Below you can find what do the preprocessing function for each net evaluated:

| Net        | Preprocessing    function       | mean/std |
| ------------- |:-------------:|:-----:|
| ResNet152V2  | Tf (x /= 127.5, x -= 1.) | None/None |
| MobileNetV2   | Tf (x /= 127.5, x -= 1.) | None/None |
| DenseNet201 | Torch (x /= 255.) | [0.485, 0.456, 0.406]/[0.229, 0.224, 0.225]  |
| EfficientNetB0 | Empty | None/None |
| VGG16/19 | Caffe (RGB to BGR) | [103.939, 116.779, 123.68]/ None |
| ResNet50 | Caffe (RGB to BGR) | [103.939, 116.779, 123.68]/ None |

For the `ResNet152V2` and `MobileNetV2` the attack success using the preprocessing function, also `DenseNet201` success. The `EfficientNetB0` has no preprocessing implementation, to success I understand that the `eps` value might be in other scale than [0-1], I used the 30 as high value to success. For `VGG16/19` and `ResNet50` the results images are black after no using preprocessing function and using the `mean/std` values to `KerasClassifier` constructor. The input of these nets need to be converted to BGR, then zero-center each color channel with respect to the ImageNet, without scaling. To success I not used the preprocessing and used 30 `eps`.

So my conclusion is that for some nets the scale of the `eps` will be different and I have to change it accordingly.

**Comment by beat-buesser** [comment]
Hi @KevinRupper Thank you very much for the overview.

The value of `eps` has to be in the same scale as the range of input values in `x`. For every model you can provide `x` to ART either in range [0, 1] or [0, 255], or any other range, if you adapt `mean`, `std`, clip_values` and `eps` accordingly to be in the same scale.

For example, `eps=8` for `x` in range [0, 255] and `clip_Values=(0, 255) corresponds to an `eps=0.03137` (8/255) for input `x` in range [0, 1] and `clip_values=(0, 1)`.

The `mean` and `std` values have to be provided in such a way that they transform `x` into the range expected by the model as input. If the model expects BGR as input, then `x` has to be in format BGR already and `mean` and `std` have to be adapted accordingly

Following your table, for example:
- for `ResNet152V2`:
-- if `x` in range [0, 255], `mean=127.5` and `std=127.5` and `eps=8`
-- if `x` in range [0, 1], `mean=0.5` and `std=0.5` and `eps=8/255`

- for `VGG16/19`: 
-- convert `x` to BGR, `x` in range [0, 255],  `mean=[103.939, 116.779, 123.68]` and `std=[1.0, 1.0, 1.0]` and `eps=8`
-- convert `x` to BGR, `x` in range [0, 1],  `mean=[103.939/255, 116.779/255, 123.68/255]` and `std=[1.0, 1.0, 1.0]` and `eps=8/255`

We are working on providing automation for these preprocessing steps, for popular models like `tf.keras.applications` etc. in future releases of `art.preprocessing`, probably with ART 1.7.

**Comment by KevinRupper** [comment]
Hi @beat-buesser, 

All it's clear now. I'm looking forward to the 1.7 version.

Thank you for your help and support.
